gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializationException.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.core.BetaApi;
+
+/**
+ * An exception thrown when a protobuf message cannot be serialized/deserialized for REST
+ * interactions.
+ */
+@BetaApi
+public class ProtoRestSerializationException extends RuntimeException {
```

Comment:
If the proto comes from external sources, as they usually do, this should be a checked exception.

Reply:
Unfortunately I can't make it checked:
1) There is a well-known controversy around checked vs unchecked exceptions (for example: https://www.ibm.com/developerworks/library/j-jtp05254/index.html). This does not prove that checked exceptions are plain wrong, but it is definitely not something that community agrees upon easily.

2)  gax-java architecture does not foresee usage of checked exceptions on its surface. If any exception happens during a call on any of the stages, sooner or later it will be wrapped to an unchecked exception, most likely one of the defined in gax itself.  I can't change it without rewriting in a backward-incompatible way the whole gax's exception handling logic. This means that if I don't wrap a checked `IOException` into an unchecked one here, it will be wrapped somewhere up the stack before a user can see it, plus going through the trouble of propagating the checked aspect of it explicitly to that level.

4) For example, check [ApiExceptionFactory](https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/ApiExceptionFactory.java#L39) - all of the created exceptions are unchecked. 
  Many of those exceptions mean that the particular call in which they were executed must be terminated and cannot be completed successfully (i.e. there is no reasonable way to recover from the exception except starting the whole call over again). This is in alignment with Oracle's recommendation (https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html): "If a client can reasonably be expected to recover from an exception, make it a checked exception. If a client cannot do anything to recover from the exception, make it an unchecked exception". If the `ProtoRestSerializationException` is thrown the only thing a user can do is to fix something in their call and try it over again from the beginning. 

5) On practice, for this particular case, an attempt to keep the exceptions checked hits the wall of troubles almost immediately. Specifically `ProtoRestSerializer#toJson()` is called from `ProtoMessageResponseParser#parse()`, which, in its turn implements [HttpResponseParser#parse()](https://github.com/googleapis/gax-java/blob/master/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpResponseParser.java#L43) interface method. To propagate it further, we will have to add `throws IOException` to the interface method, which is a breaking change, or wrap it in an unchecked on the `HttpResponseParser#parse()` level. This is where it begins, if we propagate further if will get more and more complicated until it gets wrapped anyways or gax-java exception handling model is revisited as a whole.

Reply:
HttpResponseParser#parse() does not document that it throws any exceptions. That strongly suggests it doesn't want any to be thrown, not that it thinks throwing a runtime exception is OK. You might need to return a default value here or null instead of throwing an exception.

Reply:
Yes, the interface doew not document the thrown exceptions, but they are definitely possible by the existing implementaiton of this interface https://github.com/googleapis/gax-java/blob/master/gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageHttpResponseParser.java#L94, that calls `GsonfromJson`, which throws two exceptions: `JsonIOException, JsonSyntaxException`, both unchecked (`JsonIOException` is also unchecked, even though it is named like it extends `IOException`, while it does not).  Gson (google's libraryh) implementation seems following the no-checked exceptions philosophy, same as gax.

Reply:
Swallowing the error information by returning null in case of an exception seems worse than throwing an unchecked exception, plus the existing implementation already throws them. WDYT about just updating the interface method documentation?

Reply:
I renamed `ProtoRestSerializationException` to `RestSerializationException` (to make it less proto-specific), added it (still as an unchecked exception) to the documentation of the `ApiMessageHttpResponseParse` and in other relevant classes. 

Note, `ApiMessageHttpResponseParse` had broken documentation block, so centennially speaking it never had documentation, but rather a multi-line comment (`/* */` was used instead of `/** */`).